### PR TITLE
Laser tag custom games!

### PR DESCRIPTION
**Pull request by ShiftyRail**
### Foreword

Have you ever played laser game for your birthday when you were just a young child ?
Did you get a sweet feeling of superiority upon seeing your name at the top of the score list ?
Do you want easy to track, easy to set up, laser tag games in YOUR space station with custom home rules ?

Then YOU have no choice but to say :+1: to this PR !

### Content

This PR allows you to create and customise laser tag games by clicking on your laser tag vest.
You can customise the following parameters : 
- stun duration (default to 4 seconds) ;
- the time during which the taggun is diabled (default to 0 seconds) ;
- the fire mode (laser tag/taser tag) ;
- the mode (Free For All/teams).

You can also show the scoreboard ; it lists the players you hit, the players who hit you, your efficiency (shoots landed/shoots fired...).

Other players can join or leave a game very easily. There is as little clicks involved as possible.
_You can still player laser/taser tag without any setup if you want,_ the customisation is purely facultative.

:cl:
- rscadd: You can create and join laser tag custom games by using your laser tag vest in-hand. You can still classic laser tag without any setup if you want.

### CODE REVIEW

**Comment by ShiftyRail** [comment]
Should be all as far as code is concerned. Will test it rigorously tomorrow and then it should be good 2 go

**Comment by ShiftyRail** [comment]
Now ready for review

File: code/modules/projectiles/guns/energy/tag.dm [hunk 1]
```diff
@@ -77,7 +78,20 @@
...
+		var/obj/item/clothing/suit/tag/our_tag = vest
+		if (our_tag.my_laser_tag_game)
+			if (world.time < last_disable_time)
+				to_chat(M, "<span class='warning'>Your gun is currentlty disabled!</span>")
+				return 0
+			if ((our_tag.my_laser_tag_game.fire_mode == LT_FIREMODE_LASER) && (projectile_type != laser_projectile))
+				to_chat(M, "<span class='warning'>Your laser tag game requires to play with lasers!</span>")
+				return 0
+			if ((our_tag.my_laser_tag_game.fire_mode == LT_FIREMODE_TASER) && (projectile_type != taser_projectile))
+				to_chat(M, "<span class='warning'>Your laser tag game requires to play with tasers!</span>")
```
Comment (Kurfursten) [inline code/modules/projectiles/guns/energy/tag.dm]: requires "you" to play with tasers

File: code/modules/projectiles/guns/energy/tag.dm [hunk 2]
```diff
@@ -77,7 +78,20 @@
...
-	if(istype(get_tag_armor(M), needed_vest))
+	var/obj/item/clothing/suit/tag/vest = get_tag_armor(M)
+	if(istype(vest, needed_vest))
+		var/obj/item/clothing/suit/tag/our_tag = vest
+		if (our_tag.my_laser_tag_game)
+			if (world.time < last_disable_time)
+				to_chat(M, "<span class='warning'>Your gun is currentlty disabled!</span>")
+				return 0
+			if ((our_tag.my_laser_tag_game.fire_mode == LT_FIREMODE_LASER) && (projectile_type != laser_projectile))
+				to_chat(M, "<span class='warning'>Your laser tag game requires to play with lasers!</span>")
```
Comment (Kurfursten) [inline code/modules/projectiles/guns/energy/tag.dm]: requires you to play

File: code/modules/projectiles/guns/energy/tag.dm [hunk 3]
```diff
@@ -77,7 +78,20 @@
         makeTaser(user)
 
 /obj/item/weapon/gun/energy/tag/special_check(var/mob/living/M)
-	if(istype(get_tag_armor(M), needed_vest))
+	var/obj/item/clothing/suit/tag/vest = get_tag_armor(M)
+	if(istype(vest, needed_vest))
+		var/obj/item/clothing/suit/tag/our_tag = vest
+		if (our_tag.my_laser_tag_game)
+			if (world.time < last_disable_time)
+				to_chat(M, "<span class='warning'>Your gun is currentlty disabled!</span>")
```
Comment (Kurfursten) [inline code/modules/projectiles/guns/energy/tag.dm]: currently

File: code/modules/clothing/suits/miscellaneous.dm [hunk 4]
```diff
@@ -34,6 +48,146 @@
...
+		var/choice = input(usr, "Choose the disbale duration.", "Disable duration") as null|num
+		game.disable_time = clamp(choice, 0, 30)
+		refresh_edit_window(usr, game)
+		return
+
+	if (href_list["edition_done"])
+		usr << browse(null,"window=laser_tag_window2;size=250x250")
+		return
+
+	if (href_list["delete_game"])
```
Comment (Kurfursten) [inline code/modules/clothing/suits/miscellaneous.dm:174]: I feel like this needs some sanity to prevent href exploits (may apply to several of these)

Comment (ShiftyRail) [reply code/modules/clothing/suits/miscellaneous.dm:174]: Correct, this one doesn't check if the dude deleting it is the owner of the game.
"Regular" href exploits (using it whilst dead, using it remotely, etc) are handled by `if (..()) return`

**Pull request merged**